bot/exts/info/doc/_cog.py

```diff
@@ -0,0 +1,419 @@
+        self.scheduled_inventories = set()
+
+        self.refresh_event = asyncio.Event()
+        self.refresh_event.set()
+        self.bot.loop.create_task(self.init_refresh_inventory())
+
+    @lock("doc", COMMAND_LOCK_SINGLETON, raise_error=True)
+    async def init_refresh_inventory(self) -> None:
+        """Refresh documentation inventory on cog initialization."""
+        await self.bot.wait_until_guild_available()
+        await self.refresh_inventory()
+
+    async def update_single(self, api_package_name: str, base_url: str, package: INVENTORY_DICT) -> None:
+        """
+        Rebuild the inventory for a single package.
+
+        Where:
+            * `package_name` is the package name to use, appears in the log
+            * `base_url` is the root documentation URL for the specified package, used to build
+                absolute paths that link to specific symbols
+            * `inventory_url` is the absolute URL to the intersphinx inventory.
+        """
+        self.base_urls[api_package_name] = base_url
+
+        for group, items in package.items():
+            for symbol, relative_doc_url in items:
+
+                # e.g. get 'class' from 'py:class'
+                group_name = group.split(":")[1]
+                if (original_symbol := self.doc_symbols.get(symbol)) is not None:
+                    replaced_symbol_name = self.ensure_unique_symbol_name(
+                        api_package_name,
+                        group_name,
+                        original_symbol,
+                        symbol,
+                    )
+                    if replaced_symbol_name is not None:
+                        symbol = replaced_symbol_name
+
+                relative_url_path, _, symbol_id = relative_doc_url.partition("#")
+                # Intern fields that have shared content so we're not storing unique strings for every object
+                symbol_item = DocItem(
+                    api_package_name,
+                    sys.intern(group_name),
+                    base_url,
+                    sys.intern(relative_url_path),
+                    symbol_id
+                )
+                self.doc_symbols[symbol] = symbol_item
+                self.item_fetcher.add_item(symbol_item)
+
+        log.trace(f"Fetched inventory for {api_package_name}.")
+
+    async def update_or_reschedule_inventory(
+            self,
+            api_package_name: str,
+            base_url: str,
+            inventory_url: str
+    ) -> Optional[INVENTORY_DICT]:
+        """
+        Update the cog's inventory, or reschedule this method to execute again if the remote inventory unreachable.
+
+        The first attempt is rescheduled to execute in `FETCH_RESCHEDULE_DELAY.first` minutes, the subsequent attempts
+        in `FETCH_RESCHEDULE_DELAY.repeated` minutes.
+        """
+        package = await fetch_inventory(inventory_url)
+
+        if not package:
+            if inventory_url not in self.scheduled_inventories:
```

Comment:
I've had to redesign this a bit in e5b2b725f50f940bbe92d3ff461232ad024935c3 as I realized the method didn't actually work for the subsequent scheduled updates because of the shared task id.